Backgrounding of Notifications Domain

amydevs opened the pull request:
### Description


This PR applies backgrounding to the Notifications domain by using `TaskManager`.

The sending of notifications can be either specified as blocking, or non-blocking, and with any amount of retries.

By doing this, errors on attempts to send notifications are now caught and turned into warning logs instead for non-blocking notifications.

Methods and RPC handlers to list and cancel the pending notifications are also to be added.

### Issues Fixed

* Related https://github.com/MatrixAI/Polykey-CLI/issues/168
* Fixes #703 

### Tasks

- [x] 1. Implement non-blocking sending of notifications.
- [x] 2. Implement outbox notification listing 
- [x] 3. Implement outbox notification cancelling
- [x] 4.  Tests for outbox notifications in NotificationsManager 
- [x] 5. Tests for outbox notification RPC handlers

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

amydevs committed:
feat: `NotificationsManager.sendNotification` is no longer blocking

fix: exponential backoff for notifications

fix: outbox notifications are now given ids

fix: notification tasks are now no longer lazy

fix: sendNotification now returns notificationId

fix: outbox notifications now use DB state

fix: NotificationId is now included in Notification, with the id and senderId fields

fix: `NotificationsManager` `Notification`s now have different types for state in DB and in usage

fix: tests

amydevs committed:
feat: `notificationsOutboxClear` RPC handler and caller

amydevs committed:
fix: api for `NotificationsManager.sendNotification` now takes an object

amydevs committed:
fix: `NotificationsManager.sendNotification` now returns a `sendProm` rather than blocking directly from the returned promise

fix: `NotificationsManager` tests

amydevs committed:
feat: `notificationsOutboxRead` RPC caller and handler

tegefaulkes commented on src/client/handlers/NotificationsOutboxRead.ts:
```diff
@@ -0,0 +1,43 @@
+        notification: notification,
+      };
+    }
+  }
```
Generally we want to avoid large collections like this in array form. Ideally the `readOutboxNotifications` should be a generator.

amydevs replied:
should readNotifications also be made an asyncgenerator?

tegefaulkes replied:
Yeah, anything that can return a potentially unbounded array should be a generator. Doubly so if it gets streamed over RPC.

amydevs replied:
this has been done

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -73,37 +94,174 @@ class NotificationsManager {
+  protected notificationsManagerOutboxNotificationsCounterDbPath: KeyPath = [
+    ...this.notificationsManagerOutboxDbPath,
+    NOTIFICATIONS_COUNT_KEY,
+  ];
```
Normally we include a description of what the `LevelPath` looks like. You can look at the `GestaltGraph` to see how. E.G.

```ts
 /**
   * Gestalt links.
   * `GestaltGraph/links/{GestaltLinkId} -> {json(GestaltLink)}`
   */
  public readonly dbLinksPath: Readonly<LevelPath> = [
    this.constructor.name,
    'links',
  ];

  /**
   * Node information
   * `GestaltGraph/nodes/{GestaltKey} -> {json(GestaltNodeInfo)}`
   */
  public readonly dbNodesPath: Readonly<LevelPath> = [
    this.constructor.name,
    'nodes',
  ];
```

amydevs replied:
this has been done

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -73,37 +94,174 @@ class NotificationsManager {
+  protected outboxNotificationIdEncodedToTaskIdMap: Map<
+    NotificationIdEncoded,
+    TaskId
+  >;
```
I'm not sure I like the in memory map for this. I think it's possible to use the task path to handle this. Let me look into it more.

tegefaulkes replied:
Yeah, you can make the path of the scheduled task be `[this.sendNotificationHandlerId, notificationIdEncoded]` and that will maintain the `task -> notification` relationship`. To find the `notificationId` of a task you just look at the path. To get the reverse you just need to iterate over the path with that `notificationId`. The iteration of the task path is annoying, yes, but also fine for this application. The we don't maintain a separate memory map for this information reducing the complexity.

You can make a protected utility function to get the `taskId` -> `notificationId` if you need it in a bunch of places.

amydevs replied:
this has been done, i dont ever rlly need to reference the taskid the other way around so it's alright

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -73,37 +94,174 @@ class NotificationsManager {
 
-  protected notificationIdGenerator: () => NotificationId;
+  public readonly sendNotificationHandlerId =
+    `${this.constructor.name}.sendNotificationHandler` as TaskHandlerId;
```
I've always defined this after the handler arrow function. Let's keep it consistent.

amydevs replied:
this has been done

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -73,37 +94,174 @@ class NotificationsManager {
+    )!;
+    const notificationKeyPath = [
+      ...this.notificationsManagerOutboxNotificationsDbPath,
+      idUtils.toBuffer(notificationId),
```
You don't need to use `idUtils.toBuffer()`. Any `Id` type has `toBuffer()` built in. So to keep consistent `notificationId.toBuffer()` is preferred here.

amydevs replied:
this has been done

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -73,37 +94,174 @@ class NotificationsManager {
+
+    const notificationDb = await this.db.withTransactionF(
+      async (tran) => (await tran.get<NotificationDB>(notificationKeyPath))!,
+    );
```
This is odd. If it's just a single thing in the transaction then just use `this.db.get()` instead. But I can see another transaction within a `finally` block further down. Doesn't that need to be consistent with this `get`? If so then pretty much the whole handler needs to be wrapped within a transaction.

amydevs replied:
i've encapsulated the entirety of sendNotificationHandler with the db transaction

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -73,37 +94,174 @@ class NotificationsManager {
+            notificationDb.data.type
+          } notification to ${nodesUtils.encodeNodeId(nodeId)}`,
+        );
+      }
```
`ErrorPolykeyRemote` isn't a connection related error, wraps errors thrown by the other side of the connection. A connection needs to be established first for that to happen. To check for a connection error we have a utility at `isConnectionError` `src/nodes/utils.ts:317` to check if an error is a failure to connect in some fashion.

amydevs replied:
i'm now using isConnectionError to check instead now

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -73,37 +94,174 @@ class NotificationsManager {
+          ],
+          delay: 1000 * 2 ** attempt,
+          lazy: false,
+        });
```
I don't like the unbounded delay here. Also a 1 second retry seems to short. I don't see connection issues resolving within that time. Maybe start at 1 hour and limit it to 1 day delay? That'll cap out after 5 retries. How many attempts would we normally expect?

amydevs replied:
i have made it so that the delay starts at 1 hr and is capped at 1 day, doubling each retry. The default retries is now 8.

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -73,37 +94,174 @@ class NotificationsManager {
+          );
+        }
+      });
+    }
```
The logic flow of this whole try-catch-finally block seems a little hard to follow. So the logic is

1. Send succeeds - Preform any needed cleanup in database and state indicating success
2. Send fails and out of attempts - Clean up state indicating no more retries.
3. Send fails and more attempts - schedule new attempt and modify any needed state.

Right now this is spread across the try-catch-finally block. which is why it's confusing. I think we can flatten this out a little in just the try-catch sections. Doing a state update before returning or throwing.

amydevs replied:
it's been made much simpler now with the removal of the NotificationID -> TaskID map

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -73,37 +94,174 @@ class NotificationsManager {
+      nodeIdEncoded: NodeIdEncoded;
+      notificationIdEncoded: NotificationIdEncoded;
+      retries: number;
+      attempt: number;
```
Unless we really need to know the total attempts that were allowed. We only really need to count down the attempts left and track 1 number for the logic.

So rather than count up till the retry limit, start at the retry limit and count to 0.

amydevs replied:
i tried to do that originally, but the attempts are needed to keep track of the exponential timeout

tegefaulkes replied:
You can just double the previous delay. Unless you can't access the task info inside of the handler?

amydevs replied:
i've made it so that the delay is doubled instead

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -161,26 +353,210 @@ class NotificationsManager {
+  ): Promise<{
+    notificationId: NotificationId;
+    sendProm: Promise<void>;
+  }> {
```
Probably call `sendProm` something like `sendP` or just `p`. We're avoiding using the `prom` name now.

amydevs replied:
this has been done

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -161,26 +353,210 @@ class NotificationsManager {
+      nodeId,
+      data,
+      retries = 64,
+    }: {
```
Avoid magic numbers, If we need a default here then take it from `this.defaultRetries` and set that in the constructor and `createX`. I'm not sure if we need to go as far as adding the default value for this to the config. so it should be fine to have the magic number specified in the `createX` configuration.

amydevs replied:
this has been done.

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -161,26 +353,210 @@ class NotificationsManager {
+      await this.db.withTransactionF(transactionCallback);
+    } else {
+      await transactionCallback(tran);
+    }
```
We use recursion to ensure a transaction is created when it's not provided. Do something like this to the top of a method.

```ts
  public async setNode(
    nodeInfo: GestaltNodeInfo,
    tran?: DBTransaction,
  ): Promise<['node', NodeId]> {
    if (tran == null) {
      return this.db.withTransactionF((tran) => this.setNode(nodeInfo, tran));
    }
    //...
  }
```

amydevs replied:
i've updated this to follow the suggestion

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -161,26 +353,210 @@ class NotificationsManager {
+      notificationIds.push(notificationId);
+    }
+    return notificationIds;
+  }
```
This should be a generator. Generally we want to avoid collecting generator outputs into an array unless we have to.

amydevs replied:
this is partially done, I'm still fixing the tests for this.

amydevs replied:
this has been done

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -161,26 +353,210 @@ class NotificationsManager {
+    number?: number | 'all';
+    order?: 'newest' | 'oldest';
+    tran?: DBTransaction;
+  } = {}): Promise<Array<Notification>> {
```
Likewise, this should be a generator.

amydevs replied:
should i change the readNotifications fn to also be an asyncgenerator?

tegefaulkes replied:
Yes.

tegefaulkes commented on src/notifications/NotificationsManager.ts:
```diff
@@ -73,37 +94,174 @@ class NotificationsManager {
+  protected notificationsManagerInboxNotificationsCounterDbPath: KeyPath = [
+    ...this.notificationsManagerInboxDbPath,
+    NOTIFICATIONS_COUNT_KEY,
+  ];
```
Is this a counter for the number of inbox messages? Don't do this because it WILL conflict without locking. Even with locking it's not a good idea. We already have a way to track the number of entries in the DB. there's a `tran.count` method that should handle this. See if you can try using that instead.

amydevs replied:
the inbox count already existed before i worked on the notifications, i just changed the db path

amydevs replied:
i've removed this for now

tegefaulkes reviewed (commented):
@CMCDragonkai Asked me to have a look over. Seems some things can be simplified and other things to be avoided. But otherwise looking fine.

I'll look deeper into the tests later.

amydevs committed:
fix: `NotificationsManager` outbox send logic simplification

fix: removed idUtils toBuffer usage

fix: changed position of sendNotificationHandlerId

fix: notificationsManager sendProm name

fix: removed count from NotificationsManager

fix: simplified NotificationId to TaskId reference

fix: sendNotificationHandler logic

fix: NotificationsManager db levelpath docblocks

fix: defaultRetries on NotifciationsManager constructor

amydevs committed:
feat: changed `readNotifications` and `readOutboxNotifications` to return
AsyncGenerator

fix: removed unneeded creation of transaction in private methods

fix: notificationsRead handler now uses asyncgenerator

fix: readded blocking to NotificationsSend handler

fix: tests and exports for notifications outbox handlers

amydevs committed:
fix: `NotificationsManager.sendNotification` delay is now double rather than exponential

amydevs committed:
feat: `NotificationsManager.removeNotification` and
`NotificationsRemove` RPC handlers

amydevs committed:
fix: renamed `notificationsRead`, `notificationsClear` and `notificationsRemove` RPC calls to `notificationsInboxRead`, `notificationsInboxClear` and `notificationsInboxRemove` respectively

amydevs committed:
fix: default `sendRetryIntervalTimeMin` and `sendRetryIntervalTimeMax` values + notifcation send retry tests

amydevs committed:
fix: `NotificationsManager` sendHandler doesn't need transaction block

amydevs committed:
feat: added taskMetadata to `notificationsOutboxRead` rpc call

fix: `notifications.test.ts` for RPC handlers involving notification IDs

tegefaulkes commented on src/client/handlers/NotificationsOutboxRead.ts:
```diff
@@ -0,0 +1,58 @@
+        number: input.number,
+        order: input.order,
+        tran,
+      }),
```
Technically we'd support seeking, just as a standard way of handling streaming like this. We always have seek, order, limit.

 But would seek even be useful here? Would we want to list all notifications before or after a certain time?

CMCDragonkai replied:
If it can support it, it's better to be consistently supported.

amydevs replied:
this has been done

tegefaulkes reviewed (commented):
Seems good.

amydevs committed:
feat: standardization of `NotificationsManager` method params to be more
like `Audit` domain

amydevs committed:
feat: added `seek` and `seekEnd` to `NotificationsManager`
